File: sach/de3395.cpp
//DE3395.CPP - Tran Huu Nam - hsgtin.vn - 2024
#include <bits/stdc++.h>
using namespace std;
int n;
int x[100];
int p[]={2,3,5,7,11,13,17,19,23,29,31};
long long kq=1e18;
long long luythua(int a, int n) {
    long long t=1;
    for (int i=1; i<=n; i++) t*=a;
    return t;
}
void xuli(int k) {
    long long s=1;
    for (int i=1; i<k; i++) {
        s*=luythua(p[i-1], x[i]-1);
    }
    if (s<kq) kq=s;
}
void thu(int k, int n, int g) {
    if (n==1) {
        xuli(k);
        return ;
    }
    for (int v=2; v<=min(n,g); v++) {
        if (n % v==0) {
            x[k]=v;
            thu(k+1, n/v, v);
        }
    }
}
int main()
{
    cin>>n;
    thu(1, n, n);
    cout<<kq;
    return 0;
}

File: sach/de3395.py
//DE3395.PY - Tran Huu Nam - hsgtin.vn - 2024
import math
n = 0
x = [0] * 100
p = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31]
kq = 1e18
def luythua(a, n):
    t = 1
    for i in range(1, n + 1):
        t *= a
    return t
def xuli(k):
    s = 1
    for i in range(1, k):
        s *= luythua(p[i - 1], x[i] - 1)
    global kq
    if s < kq:
        kq = s
def thu(k, n, g):
    if n == 1:
        xuli(k)
        return
    for v in range(2, min(n, g) + 1):
        if n % v == 0:
            x[k] = v
            thu(k + 1, n // v, v)
if __name__ == "__main__":
    n = int(input())
    thu(1, n, n)
    print(kq)

